lib/puppet/util/yaml.rb

```diff
@@ -9,6 +9,48 @@ module Puppet::Util::Yaml
 
   class YamlLoadError < Puppet::Error; end
 
+  def self.disallowed_class(cls)
+    raise YamlLoadError, "Tried to load unspecified class: #{cls}"
+  end
+
+  def self.safe_load(yaml, allowed_classes = [], filename = nil)
+    @psych_safe_load ||= Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('2.1.0')
+    return YAML.safe_load(yaml, allowed_classes, [], false, filename) if @psych_safe_load
```

Comment:
I think it'd be clearer to write the methods differently, and then you don't need a class variable:

```ruby
if Gem::Version.new(RUBY_VERSION.dup) >= Gem::Version.new('2.1.0')
  def self.safe_load(yaml, allowed_classes = [], filename = nil)
    YAML.safe_load(yaml, allowed_classes, [], false, filename)
  end
else
   def self.safe_load(yaml, allowed_classes = [], filename = nil)
     # ruby 1.9/2.0 implementation
  end
end
```

Note some older rubies have trouble with `Gem::Version.new(RUBY_VERSION)` which is why you'll often see `RUBY_VERSION.dup` used.

Reply:
The problem with `Gem::Version#new` and frozen strings was [fixed in Gem version 2.0.0](https://github.com/rubygems/rubygems/commit/da4362a6644ca5a75c210677ac500bccfe75f529#diff-94d386526b8b558035d4e4c7602d647c) (5 years ago). Do we still expect to run into that?

Reply:
MRI ruby 1.9.3 includes rubygems 1.8.23.2, which [doesn't have that fix](https://github.com/rubygems/rubygems/blob/v1.8.23.2/lib/rubygems/version.rb#L190). See also https://github.com/puppetlabs/puppet/commit/fb9995da521269c17755c15dcb2d19afa7932501.